### PR TITLE
Make usage of absolute links configurable

### PR DESCRIPTION
**Pull request by sebastian-meyer**
In multi-domain environments like the Zeitungsportal we need the option to force all links to be absolute. Since TYPO3 9.x the Typoscript configuration option `config.absRefPrefix` only effects links generated by Typoscript, but doesn't work globally.

This Pull-Request introduces new extension-wide configuration options to force all generated links to be absolute. In addition it allows using HTTPS as default scheme for those URLs.

The default is to not use absolute links, so existing instances should behave exactly like before.

### CODE REVIEW

File: Classes/Common/AbstractPlugin.php [hunk 1]
```diff
@@ -198,40 +199,68 @@ protected function loadDocument()
...
+    {
+        $parser = GeneralUtility::makeInstance(\TYPO3\CMS\Core\TypoScript\Parser\TypoScriptParser::class);
+        $parser->parse($string);
+        return $parser->setup;
+    }
+
+    /**
+     * Link string to the current page.
+     * @see \TYPO3\CMS\Frontend\Plugin\AbstractPlugin->pi_linkTP()
+     *
```
Comment (stweil) [inline Classes/Common/AbstractPlugin.php]:
```suggestion
     *
```

Comment (stweil) [reply Classes/Common/AbstractPlugin.php]: This fixes Codacy CI.

Comment (sebastian-meyer) [reply Classes/Common/AbstractPlugin.php]: Thanks!

**Review by albig — DISMISSED** [review]
Actually this seems to work but keep in mind:
With TYPO9.5: 
* You need a site-configuration with a base including scheme+domain. Without this the links keep to be relative. Providing "/" as base will change nothing with these settings and won't make absolute URLs.
* If you have no HTTPS scheme provided in the site-configuration, it won't be force by this setting.
* Don't know, what is happening with TYPO3 8.7.

As it shouldn't break anything, we can merge this even if it won't work as expected in all situaltions.

**Comment by sebastian-meyer** [comment]
That's how it is intended to work:
- If you have configured a base URL and HTTPS scheme in the `site-configuration`, those do not automatically lead to TYPO3 generating absolute URLs everywhere. The `typolink` function will still generate relative links as long as you don't set the `forceAbsoluteUrl` parameter. So that's where the new extension settings come into play.
- If you haven't configured a base URL and/or HTTPS scheme in the `site-configuration`, you can't expect TYPO3 to generate absolute URLs even with the new settings.

But you are right, I should make it clearer, that having a proper `site-configuration` is required for using the new extension settings.

**Review by albig — APPROVED** [review]

File: Classes/Common/AbstractPlugin.php [hunk 2]
```diff
@@ -198,40 +199,68 @@ protected function loadDocument()
...
-            if (!$GLOBALS['TSFE']->config['config']['disablePrefixComment']) {
-                $content = "\n\n<!-- BEGIN: Content of extension '" . $this->extKey . "', plugin '" . Helper::getUnqualifiedClassName(get_class($this)) . "' -->\n\n" . $content . "\n\n<!-- END: Content of extension '" . $this->extKey . "', plugin '" . Helper::getUnqualifiedClassName(get_class($this)) . "' -->\n\n";
-            }
+        $conf = [];
+        if (!$cache) {
+            $conf['no_cache'] = true;
         }
-        return $content;
+        $conf['parameter'] = $altPageId ?: ($this->pi_tmpPageId ?: 'current');
+        $conf['additionalParams'] = $this->conf['parent.']['addParams'] . HttpUtility::buildQueryString($urlParameters, '&', true) . $this->pi_moreParams;
```
Comment (albig) [inline Classes/Common/AbstractPlugin.php:237]: buildQueryString() is unknown in TYPO3 8.7.

Comment (sebastian-meyer) [reply Classes/Common/AbstractPlugin.php:237]: Good catch! I'll fix it in a separate pull request.

**Pull request merged**